oauth_provider_jwt/models/oauth_provider_client.py

```diff
@@ -0,0 +1,187 @@
+        ('PS256', 'RSASSA-PSS signature using SHA-256 and MGF1 padding with '
+         'SHA-256'),
+        ('PS384', 'RSASSA-PSS signature using SHA-384 and MGF1 padding with '
+         'SHA-384'),
+        ('PS512', 'RSASSA-PSS signature using SHA-512 and MGF1 padding with '
+         'SHA-512'),
+    ], string='Algorithm', help='Algorithm used to sign the JSON Web Token.')
+    jwt_private_key = fields.Text(
+        string='Private Key',
+        help='Private key used for the JSON Web Token generation.')
+    jwt_public_key = fields.Text(
+        string='Public Key', compute='_compute_jwt_public_key',
+        help='Public key used for the JSON Web Token generation.')
+
+    @api.multi
+    def _load_private_key(self):
+        """ Load the client's private key into a cryptography's object instance
+        """
+        return load_pem_private_key(
+            str(self.jwt_private_key),
+            password=None,
+            backend=default_backend(),
+        )
+
+    @api.multi
+    @api.constrains('jwt_algorithm', 'jwt_private_key')
+    def _check_jwt_private_key(self):
+        """ Check if the private key's type matches the selected algorithm
+
+        This check is only performed for asymetric algorithms
+        """
+        for client in self:
+            algorithm_prefix = client.jwt_algorithm[:2]
+            if algorithm_prefix in ALGORITHMS and client.jwt_private_key:
+                private_key = client._load_private_key()
+
+                if not isinstance(
+                   private_key, ALGORITHMS[algorithm_prefix]):
+                    raise exceptions.ValidationError(
+                        _('The private key doesn\'t fit the selected '
+                          'algorithm!'))
+
+    @api.multi
+    def generate_private_key(self):
+        """ Generate a private key for ECDSA and RSA algorithm clients """
+        for client in self:
+            algorithm_prefix = client.jwt_algorithm[:2]
+            if algorithm_prefix not in ALGORITHMS:
+                continue
+
+            if algorithm_prefix == 'ES':
+                key = ec.generate_private_key(
+                    curve=ec.SECT283R1,
+                    backend=default_backend(),
+                )
+            elif algorithm_prefix in ('RS', 'PS'):
+                key = rsa.generate_private_key(
+                    public_exponent=65537, key_size=2048,
+                    backend=default_backend(),
+                )
+
+            client.jwt_private_key = key.private_bytes(
+                encoding=Encoding.PEM,
+                format=PrivateFormat.TraditionalOpenSSL,
+                encryption_algorithm=NoEncryption(),
+            )
+
+    @api.multi
+    def _compute_jwt_public_key(self):
+        """ Compute the public key associated to the client's private key
+
+        This is only done for asymetric algorithms
+        """
+        for client in self:
+            if client.jwt_algorithm.startswith(('ES', 'RS', 'PS')) and \
```

Comment:
IMO all the hard coded `ES`, `RS`, `PS` somewhat defeats the purpose of having the constant. What about:

```
if client.jwt_private_key and client.jwt_algorithm[:2] in ALGORITHMS:
```

Also note the boolean eval before the `in` eval - it's cheaper, so might as well knock any of those out first.